executor: support merge join on disk

ichn-hu opened the pull request:
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/community/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

This PR enables merge join to spill rows with same key to disk so that the tidb server won't go OOM if there are two many rows with the same key.

### What is changed and how it works?

Created a `RowContainer` that handles rows in memory and also in disk.

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test

Code changes

No user visible changes, however internally, `MergeJoinExec` and `innerJoinTable` has been changed.

Side effects

 - Possible performance regression, the regression can be minor, see benchmark below.
 - Increased code complexity.

Related changes

 - Need to cherry-pick to the release branch, should be included in release 4.0.

Release note

- We now support spilling to disk during merge join if there are some key to be joined has too many rows that surpass the capacity of the memory.

ichn-hu commented:
/run-all-tests

ichn-hu commented:
/run-all-tests

SunRunAway commented on executor/merge_join.go:
```diff
@@ -109,16 +109,20 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
 	t.resourceQueue = append(t.resourceQueue, t.resultQueue[0:lastResultIdx]...)
 	t.resultQueue = t.resultQueue[lastResultIdx:]
+	li := chunk.NewList(t.reader.base().retFieldTypes, t.curResult.Capacity(), t.curResult.Capacity())
+	rc := chunk.NewRowContainer(li, nil)
```
How to trigger spilling?
And we should add unit test for spilling.

ichn-hu replied:
Spilling test is added in row_container_test.go, and the trigger is now registered at rowsWithSameKeys where a new RowContainer is allocated.

SunRunAway commented:
It is preferable to have a benchmark to show that it does not have performace regression when replacing to `rowContainer`

ichn-hu commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,182 @@
+
+// NewRowContainer creates a new RowContainer, note that initialList
+// and initialListInDisk must have one and only one none-nil
+func NewRowContainer(initialList *List, initialListInDisk *ListInDisk) RowContainer {
```
Removed

ichn-hu commented on executor/merge_join.go:
```diff
@@ -109,16 +109,20 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
 	lastResultIdx := len(t.resultQueue) - 1
 	t.resourceQueue = append(t.resourceQueue, t.resultQueue[0:lastResultIdx]...)
 	t.resultQueue = t.resultQueue[lastResultIdx:]
+	li := chunk.NewList(t.reader.base().retFieldTypes, t.curResult.Capacity(), t.curResult.Capacity())
```
You mean reuse right? It is now reused if it is not spilled, see the updated code.

SunRunAway commented on util/chunk/list.go:
```diff
@@ -145,6 +155,15 @@ func (l *List) Reset() {
+func (l *List) Clear() {
+	l.memTracker.Consume(-l.memTracker.BytesConsumed())
+	l.freelist = l.freelist[:0]
+	l.chunks = l.chunks[:0]
```
Slice cutting does not release the reference of underlying array, so it does not free the memory actually.

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+// NewRowContainerFromList creates a new RowContainer from chunk.List
+func NewRowContainerFromList(li *List) RowContainer {
```
Why we need this function?

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,228 @@
+			return
+		}
+	}
+	c.records.Clear()
```
Why not just set `c.records` to nil?

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+// NewRowContainerFromListInDisk creates a new RowContainer from chunk.ListInDisk
+func NewRowContainerFromListInDisk(li *ListInDisk) RowContainer {
```
Why we need this function?
For unit test, you can just modify the entries in `RowContainer`.

ichn-hu replied:
The reason that I did not set `c.records` to nil is because that looses the memTracker of c.records that is setup when the chunk.List is created. If we set it to nil, then next time when we call reset, it can be tricky to setup the memTracker.

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+// Reset resets RowContainer
+func (c *RowContainer) Reset() error {
```
Why we need this method?

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,227 @@
+			if err != nil {
+				return err
+			}
+			c.records.Clear() // GC chunks in memory
```
```suggestion
			c.records = nil // GC chunks in memory
```

SunRunAway replied:
You've called `c.records.Clear()` both here and in `c.spillToDisk()`.

SunRunAway commented on util/chunk/disk.go:
```diff
@@ -83,11 +85,24 @@ func NewListInDisk(fieldTypes []*types.FieldType) *ListInDisk {
 
+// Len returns the number of rows in ListInDisk
+func (l *ListInDisk) Len() int {
+	return l.numRowsInDisk
```
Could we count the `l.offsets` to get the total number rows?

ichn-hu replied:
No, l.offsets are a two dimensional array, you can't count the number of rows unless you iterate through the first dimension, however doing so introduces complexity concern, turning a O(1) operation into O(number of chunks), therefore I prefer to keep it.

SunRunAway commented on util/chunk/disk.go:
```diff
@@ -109,15 +124,47 @@ func (l *ListInDisk) Add(chk *Chunk) (err error) {
+			return
+		}
+		l.bufWriter = bufWriterPool.Get().(*bufio.Writer)
+		l.bufWriter.Reset(l.disk)
```
We can make a method called `initDiskFile` for callings of `AppendRow` and `Add` to avoiding copying code.

ichn-hu replied:
Good idea, will do.

SunRunAway commented on util/chunk/disk_test.go:
```diff
@@ -83,6 +83,51 @@ func (s *testChunkSuite) TestListInDisk(c *check.C) {
+	it2 := NewIterator4List(records)
+	for row, row2 := it.Begin(), it2.Begin(); row != it.End() && row2 != it2.End(); row, row2 = it.Next(), it2.Next() {
+		c.Assert(row.GetDatumRow(fields), check.DeepEquals, row2.GetDatumRow(fields))
+	}
```
What is this test for?

ichn-hu replied:
All right, it is redundant, since the iterator has already be tested in iterator_test.go, I'll consider remove it.

SunRunAway commented on executor/merge_join.go:
```diff
@@ -109,16 +111,31 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
 	t.resourceQueue = append(t.resourceQueue, t.resultQueue[0:lastResultIdx]...)
 	t.resultQueue = t.resultQueue[lastResultIdx:]
+	if t.sameKeyRows == nil {
+		rc := chunk.NewRowContainer(t.reader.base().retFieldTypes, t.curResult.Capacity())
```
Would it be better that `chunk.NewRowContainer` returns a pointer of `RowContainer`?

ichn-hu replied:
Yes, will do.

SunRunAway commented on executor/merge_join.go:
```diff
@@ -109,16 +111,31 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
+	if config.GetGlobalConfig().OOMUseTmpStorage {
+		actionSpill := t.sameKeyRows.ActionSpill()
+		ctx.GetSessionVars().StmtCtx.MemTracker.FallbackOldAndSetNewAction(actionSpill)
+	}
```
Since we've reused `rowContainer`, could we just use on action?

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,227 @@
+	fallbackAction memory.ActionOnExceed
+}
+
+// Action sends a signal to trigger spillToDisk method of hashRowContainer
```
```suggestion
// Action sends a signal to trigger spillToDisk method of RowContainer
```

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,227 @@
+			a.fallbackAction.Action(t)
+		}
+	}
+	a.once.Do(func() {
```
We should make sure that the action can be triggered again after `RowContainer.Reset` is called.
And please add a unit test for this situation.

SunRunAway commented on util/chunk/row_container_test.go:
```diff
@@ -0,0 +1,113 @@
+	c.Assert(err, check.IsNil)
+
+	// renew the action
+	tracker.FallbackOldAndSetNewAction(rc.ActionSpill())
```
IMO, the action which we created before the calling of `rc.Reset()` should be able to trigger again after the calling of `rc.Reset()`, so the test code should be,

```go
action := rc.ActionSpill()
// test the `action`
rc.Reset()
// test the `action` again
```

ichn-hu replied:
It is indeed preferable if we could do so, however the spill action is registered on the Memtracker of the StmtCtx, it seems impossible to reset it inside row container without some dirty hack. Do you have any suggestions?

ichn-hu replied:
I suppose I could pass the StmtCtx or sessionctx into RowContainer, however I think it is somehow overkill and it can be harder to write unit test since during testing we don't always have the ctx.

SunRunAway commented on executor/merge_join.go:
```diff
@@ -128,7 +145,10 @@ func (t *mergeJoinInnerTable) rowsWithSameKey() ([]chunk.Row, error) {
-			t.sameKeyRows = append(t.sameKeyRows, selectedRow)
+			_, err := t.sameKeyRows.AppendRow(selectedRow)
+			if err != nil {
+				return t.sameKeyRows, err
```
```suggestion
				return nil, err
```

ichn-hu replied:
Done.

SunRunAway commented on executor/merge_join.go:
```diff
@@ -46,7 +48,7 @@ type MergeJoinExec struct {
 
-	innerRows     []chunk.Row
+	innerRows     *chunk.RowContainer
 	innerIter4Row chunk.Iterator
```
You should check `innerIter4Row.Error()` somewhere.

ichn-hu replied:
It is checked inside joiner.go.

SunRunAway commented on util/chunk/iterator.go:
```diff
@@ -284,3 +303,72 @@ func (it *iterator4RowPtr) ReachEnd() {
+	if it.rowIdx < 0 || it.chkIdx >= it.c.NumChunks() {
+		return it.End()
+	}
+	row, err := it.c.GetRow(RowPtr{uint32(it.chkIdx), uint32(it.rowIdx)})
```
Because it may read row from disk, should we cache the current row in memory?
So that we can avoid the extra disk read or deserialization for the multiple calls of `Current` in merge_join.go.

I've also noticed that `rowInDisk` has its own `bufReader`, we may do a benchmark to choose whether the cache of current row is necessary.

ichn-hu replied:
Orz

ichn-hu commented on executor/hash_table.go:
```diff
@@ -208,22 +174,20 @@ func (c *hashRowContainer) PutChunkSelected(chk *chunk.Chunk, selected []bool) e
-		chkIdx = uint32(c.recordsInDisk.NumChunks())
-		err := c.recordsInDisk.Add(chk)
+		chkIdx = uint32(c.rowContainer.NumChunks())
+		err := c.rowContainer.Add(chk)
```
Sounds good, man you are a genius!

ichn-hu commented:
The benchmark result:

## with master

![bench-master-flame](https://user-images.githubusercontent.com/29735669/71460224-f33bd480-27e5-11ea-956a-8b22bdca1bfd.png)

```
(pprof) top 20
Showing nodes accounting for 78.84s, 93.45% of 84.37s total
Dropped 229 nodes (cum <= 0.42s)
Showing top 20 nodes out of 54
      flat  flat%   sum%        cum   cum%
    64.42s 76.35% 76.35%     64.42s 76.35%  runtime.memmove
     3.38s  4.01% 80.36%     35.82s 42.46%  github.com/pingcap/tidb/util/chunk.(*Chunk).AppendPartialRow
     2.88s  3.41% 83.77%      2.88s  3.41%  runtime.madvise
     2.09s  2.48% 86.25%      2.09s  2.48%  runtime.memclrNoHeapPointers
     1.30s  1.54% 87.79%      1.30s  1.54%  github.com/pingcap/tidb/util/chunk.(*Column).IsNull
     1.06s  1.26% 89.05%      1.06s  1.26%  github.com/pingcap/tidb/util/chunk.(*Column).appendNullBitmap
     0.62s  0.73% 89.78%     37.49s 44.44%  github.com/pingcap/tidb/executor.(*innerJoiner).tryToMatchInners
     0.50s  0.59% 90.38%      0.50s  0.59%  runtime.futex
     0.48s  0.57% 90.94%     41.79s 49.53%  github.com/pingcap/tidb/executor.(*MergeJoinExec).joinToChunk
     0.36s  0.43% 91.37%      1.25s  1.48%  github.com/pingcap/tidb/executor.(*MergeJoinExec).compare
     0.29s  0.34% 91.72%      1.04s  1.23%  runtime.scanobject
     0.24s  0.28% 92.00%      1.41s  1.67%  github.com/pingcap/tidb/executor.(*mergeJoinInnerTable).rowsWithSameKey
     0.18s  0.21% 92.21%     36.18s 42.88%  github.com/pingcap/tidb/executor.(*baseJoiner).makeJoinRowToChunk
     0.18s  0.21% 92.43%      0.87s  1.03%  github.com/pingcap/tidb/util/chunk.Row.IsNull
     0.18s  0.21% 92.64%      2.03s  2.41%  runtime.mallocgc
     0.16s  0.19% 92.83%      0.50s  0.59%  github.com/pingcap/tidb/expression.(*Column).EvalInt
     0.15s  0.18% 93.01%      0.65s  0.77%  github.com/pingcap/tidb/executor.compareChunkRow
     0.15s  0.18% 93.18%      0.66s  0.78%  github.com/pingcap/tidb/expression.CompareInt
     0.12s  0.14% 93.33%      7.10s  8.42%  github.com/pingcap/tidb/executor.buildMockDataSource
     0.10s  0.12% 93.45%     14.80s 17.54%  github.com/pingcap/tidb/util/chunk.(*Chunk).AppendRow

```

## with rowContainer

![bench-pr-flame](https://user-images.githubusercontent.com/29735669/71460229-f6cf5b80-27e5-11ea-831e-9588e821bcb5.png)

```
(pprof) top 20
Showing nodes accounting for 75.73s, 91.27% of 82.97s total
Dropped 246 nodes (cum <= 0.41s)
Showing top 20 nodes out of 75
      flat  flat%   sum%        cum   cum%
    60.34s 72.73% 72.73%     60.34s 72.73%  runtime.memmove
     2.79s  3.36% 76.09%     32.79s 39.52%  github.com/pingcap/tidb/util/chunk.(*Chunk).AppendPartialRow
     2.71s  3.27% 79.35%      2.71s  3.27%  runtime.memclrNoHeapPointers
     2.33s  2.81% 82.16%      2.33s  2.81%  runtime.madvise
     1.23s  1.48% 83.64%      1.23s  1.48%  github.com/pingcap/tidb/util/chunk.(*Column).IsNull
     0.99s  1.19% 84.84%      0.99s  1.19%  github.com/pingcap/tidb/util/chunk.(*Column).appendNullBitmap
     0.69s  0.83% 85.67%     22.99s 27.71%  github.com/pingcap/tidb/executor.(*innerJoiner).tryToMatchInners
     0.59s  0.71% 86.38%     16.15s 19.46%  github.com/pingcap/tidb/executor.(*mergeJoinInnerTable).rowsWithSameKey
     0.56s  0.67% 87.06%      0.56s  0.67%  github.com/pingcap/tidb/util/memory.(*Tracker).Consume
     0.52s  0.63% 87.68%      0.53s  0.64%  github.com/pingcap/tidb/util/chunk.(*Chunk).NumRows
     0.44s  0.53% 88.21%      0.44s  0.53%  runtime.futex
     0.40s  0.48% 88.69%      2.98s  3.59%  runtime.mallocgc
     0.37s  0.45% 89.14%     41.69s 50.25%  github.com/pingcap/tidb/executor.(*MergeJoinExec).joinToChunk
     0.32s  0.39% 89.53%      1.15s  1.39%  runtime.scanobject
     0.31s  0.37% 89.90%      0.43s  0.52%  runtime.heapBitsSetType
     0.28s  0.34% 90.24%      0.65s  0.78%  github.com/pingcap/tidb/util/chunk.(*iterator4RowContainer).Current
     0.24s  0.29% 90.53%      0.76s  0.92%  github.com/pingcap/tidb/executor.compareChunkRow
     0.23s  0.28% 90.80%      1.44s  1.74%  github.com/pingcap/tidb/util/chunk.(*iterator4RowContainer).Next
     0.21s  0.25% 91.06%      0.54s  0.65%  github.com/pingcap/tidb/executor.(*mergeJoinInnerTable).nextRow
     0.18s  0.22% 91.27%      0.85s  1.02%  github.com/pingcap/tidb/util/chunk.Row.IsNull

```

ichn-hu commented:
```
benchstat master.txt disk.txt
name                                                                                                                                                old time/op    new time/op    delta
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8     558ms ± 5%     521ms ± 4%  -6.63%  (p=0.032 n=5+5)
MergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8       325ms ± 2%     324ms ± 2%    ~     (p=0.841 n=5+5)
MergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8         320ms ± 2%     330ms ± 2%  +3.36%  (p=0.032 n=5+5)

name                                                                                                                                                old alloc/op   new alloc/op   delta
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8    68.0MB ± 0%    68.0MB ± 0%    ~     (p=0.841 n=5+5)
MergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8      57.9MB ± 0%    57.9MB ± 0%    ~     (p=0.952 n=5+5)
MergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8        58.6MB ± 0%    58.6MB ± 0%    ~     (p=0.683 n=5+5)

name                                                                                                                                                old allocs/op  new allocs/op  delta
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8      304k ± 0%      304k ± 0%    ~     (p=0.992 n=5+5)
MergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8       4.11k ± 0%     4.11k ± 0%    ~     (p=0.857 n=5+5)
MergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8           764 ± 0%       764 ± 0%    ~     (p=0.643 n=5+5)
```

ichn-hu commented:
@SunRunAway @qw4990 PTAL

ichn-hu commented:
/run-all-tests

ichn-hu commented:
benchmark result:
```
(base)  ✘ ichn@ichn-arch-pc  /tmp/benchmark  benchstat master4/master.txt disk4/disk.txt
name                                                                                                                                                old time/op    new time/op    delta
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8     541ms ± 8%     613ms ± 9%       ~     (p=0.095 n=5+5)
MergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8       331ms ± 2%     341ms ± 1%     +3.08%  (p=0.016 n=5+4)
MergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8         318ms ± 4%     322ms ± 2%       ~     (p=0.421 n=5+5)
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:330000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8     586ms ±11%     605ms ± 6%       ~     (p=0.841 n=5+5)

name                                                                                                                                                old alloc/op   new alloc/op   delta
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8    68.0MB ± 0%    76.9MB ± 0%    +13.12%  (p=0.008 n=5+5)
MergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8      57.9MB ± 0%    65.9MB ± 0%    +13.72%  (p=0.008 n=5+5)
MergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8        58.6MB ± 0%    64.5MB ± 0%     +9.98%  (p=0.008 n=5+5)
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:330000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8    68.0MB ± 0%    76.9MB ± 0%    +13.10%  (p=0.008 n=5+5)

name                                                                                                                                                old allocs/op  new allocs/op  delta
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8      304k ± 0%      608k ± 0%    +99.94%  (p=0.016 n=4+5)
MergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8       4.11k ± 0%    33.44k ± 0%   +713.11%  (p=0.008 n=5+5)
MergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8           764 ± 0%      8543 ± 0%  +1018.46%  (p=0.008 n=5+5)
MergeJoinExec/merge_join_(outerRows:300000,_innerRows:330000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-8      304k ± 0%      608k ± 0%    +99.93%  (p=0.008 n=5+5)
```

SunRunAway commented on executor/merge_join.go:
```diff
@@ -172,36 +205,49 @@ func (t *mergeJoinInnerTable) hasNullInJoinKey(row chunk.Row) bool {
-		t.resourceQueue = append(t.resourceQueue, newChunk)
-	}
+	// Allocate a new Chunk from RowContainer
+	newChk := t.sameKeyRows.AllocChunk()
```
Why not just use `newFirstChunk`? Since `newChk` is not obtained by `sameKeyRows`.

ichn-hu replied:
Because we wanna reuse some chunks in the rowContainer. If the last time the RC has allocated some chunks, and it did not spill, then Reset the RC won't GC all these chunks, and next time AllocChunk will reuse these chunks.

SunRunAway commented on util/chunk/disk.go:
```diff
@@ -109,15 +131,45 @@ func (l *ListInDisk) Add(chk *Chunk) (err error) {
+
+// AppendRow appends a row to the last chunk in disk, if no chunk has ever been added,
+// a new chunk will be created. Note that the write will be flushed to disk every time.
+func (l *ListInDisk) AppendRow(row Row) (ptr RowPtr, err error) {
```
Is this method not used anymore?

ichn-hu replied:
Yes, I'll remove it.

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,243 @@
+	})
+}
+
+func (a *spillDiskAction) SetFallback(fallback memory.ActionOnExceed) {
```
Add comment for golint

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,243 @@
+	a.fallbackAction = fallback
+}
+
+func (a *spillDiskAction) SetLogHook(hook func(uint64)) {}
```
Add comment for golint

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,243 @@
+
+func (a *spillDiskAction) SetLogHook(hook func(uint64)) {}
+
+func (a *spillDiskAction) Reset() {
```
```suggestion
func (a *spillDiskAction) reset() {
```

SunRunAway commented on executor/merge_join.go:
```diff
@@ -109,56 +117,81 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
+func (t *mergeJoinInnerTable) selectedRows() *chunk.RowContainer {
+	if t.curSel != nil {
+		t.curChk.SetSel(t.curSel)
+		t.sameKeyRows.SetTemporary(t.curChk)
```
How about implement a multiple iterator like [`io.MultiReader`](https://golang.org/pkg/io/#MultiReader) to avoiding this confusing `SetTemporary` method?

ichn-hu replied:
Sounds good, maybe next PR....

qw4990 replied:
This method lets me feel confused too... How about adding a TODO comment here?

SunRunAway commented on executor/merge_join.go:
```diff
@@ -109,56 +117,81 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
 		if !t.hasNullInJoinKey(result) {
 			return result, nil
 		}
+		fmt.Println("has null key!")
```
~

ichn-hu commented:
/run-all-tests

SunRunAway commented on executor/merge_join_test.go:
```diff
@@ -235,6 +237,42 @@ func checkPlanAndRun(tk *testkit.TestKit, c *C, plan string, sql string) *testki
+	result = checkMergeAndRun(tk, c, "select /*+ TIDB_SMJ(t) */ * from t left outer join t1 on t.c1 = t1.c1 where t1.c1 = 3 or false")
+	result.Check(testkit.Rows())
+	result = checkMergeAndRun(tk, c, "select /*+ TIDB_SMJ(t) */ * from t left outer join t1 on t.c1 = t1.c1 and t.c1 != 1 order by t1.c1")
+	result.Check(testkit.Rows("1 1 <nil> <nil>", "2 2 2 3"))
```
Could you check diskTracker and memTracker to make sure they are zero here?

ichn-hu replied:
Done.

ichn-hu replied:
Also checked that disk.MaxConsume > 0.

qw4990 commented on util/chunk/list.go:
```diff
@@ -71,6 +71,16 @@ func (l *List) NumChunks() int {
+// NumRowsOfChunk returns the number of rows of a chunk in the ListInDisk.
+func (l *List) NumRowsOfChunk(chkID int) int {
+	return l.chunks[chkID].NumRows()
+}
```
This method seems redundant since we can use `List.GetChunk(chkID).NumRows()`.

ichn-hu replied:
You are right, however `ListInDisk` implemented NumRowsOfChunk and can't be removed because there is no `ListInDisk.GetChunk`, therefore I prefer to keep it for the sake of isometry. 

```
// NumRowsOfChunk returns the number of rows of a chunk in the ListInDisk.
func (c *RowContainer) NumRowsOfChunk(chkID int) int {
	if c.tmp != nil && chkID == c.NumChunks()-1 {
		return c.tmp.NumRows()
	}
	if c.AlreadySpilled() {
		return c.recordsInDisk.NumRowsOfChunk(chkID)
	}
	return c.records.NumRowsOfChunk(chkID)
}
```

qw4990 commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,246 @@
+	"github.com/pingcap/tidb/util/memory"
+	"github.com/pingcap/tidb/util/stringutil"
+
+	"go.uber.org/zap"
```
Please format these lines ~

ichn-hu replied:
Thanks, will do.

qw4990 commented on executor/merge_join.go:
```diff
@@ -109,51 +117,75 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
-	t.resultQueue = t.resultQueue[lastResultIdx:]
+func (t *mergeJoinInnerTable) selectRow(row chunk.Row) {
+	t.curSel = append(t.curSel, row.Idx())
+}
```
How about removing this method since it just has one line? 👀

ichn-hu replied:
I suppose it is more semantic and less confusing to have such a method, I prefer to keep.

qw4990 commented on executor/merge_join.go:
```diff
@@ -109,51 +117,75 @@ func (t *mergeJoinInnerTable) init(ctx context.Context, chk4Reader *chunk.Chunk)
 		compareResult := compareChunkRow(t.keyCmpFuncs, selectedRow, t.firstRow4Key, t.joinKeys, t.joinKeys)
 		if compareResult == 0 {
-			t.sameKeyRows = append(t.sameKeyRows, selectedRow)
+			t.selectRow(selectedRow)
```
How about not using the `Chunk.SetSel([]int)` method and just appending `RowPtrs` into the `RowContainer`, then the `tmp` Chunk in it can be removed, which may make it simpler?

ichn-hu replied:
Yes, that's how I implemented it before, and found severe performance regression, therefore I have to reduce the copy (runtime.memmove) by such an ugly implementation. I'll refactor this part and remove tmp using io.multireader, but AppendRow can't be used.

ichn-hu commented:
/run-all-tests

ichn-hu commented:
PTAL @SunRunAway @qw4990

SunRunAway commented on executor/merge_join_test.go:
```diff
@@ -258,19 +258,16 @@ func (s *testSuite2) TestMergeJoinInDisk(c *C) {
-	result.Check(testkit.Rows())
-	result = checkMergeAndRun(tk, c, "select /*+ TIDB_SMJ(t) */ * from t left outer join t1 on t.c1 = t1.c1 where t1.c1 = 3 or false")
-	result.Check(testkit.Rows())
-	result = checkMergeAndRun(tk, c, "select /*+ TIDB_SMJ(t) */ * from t left outer join t1 on t.c1 = t1.c1 and t.c1 != 1 order by t1.c1")
```
Why these cases are deleted?

ichn-hu replied:
Because I'm tired...

ichn-hu replied:
The first test case suffice to prove the correctness of the disk behavior, so I just removed other cases.

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,217 @@
+	return c.records.GetRow(ptr), nil
+}
+
+// GetMemTracker returns the memory tracker in records, panics if the RowContainer has already spilled
```
```suggestion
// GetMemTracker returns the memory tracker in records, panics if the RowContainer has already spilled.
```

ichn-hu replied:
...

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,217 @@
+	return c.records.GetChunk(chkIdx)
+}
+
+// GetRow returns the row the ptr pointed to
```
```suggestion
// GetRow returns the row the ptr pointed to.
```

ichn-hu replied:
My fault..

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,217 @@
+	return c.records.allocChunk()
+}
+
+// GetChunk returns chkIdx th chunk of in memory records
```
```suggestion
// GetChunk returns chkIdx th chunk of in memory records.
```

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,217 @@
+	return
+}
+
+// AllocChunk allocates a new chunk from RowContainer
```
```suggestion
// AllocChunk allocates a new chunk from RowContainer.
```

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,217 @@
+	return c.records.NumChunks()
+}
+
+// Add appends a chunk into the RowContainer
```
```suggestion
// Add appends a chunk into the RowContainer.
```

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,217 @@
+	return
+}
+
+// Reset resets RowContainer
```
```suggestion
// Reset resets RowContainer.
```

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,217 @@
+	actionSpill *spillDiskAction
+}
+
+// NewRowContainer creates a new RowContainer in memory
```
```suggestion
// NewRowContainer creates a new RowContainer in memory.
```

SunRunAway commented on util/chunk/row_container.go:
```diff
@@ -0,0 +1,217 @@
+	"go.uber.org/zap"
+)
+
+// RowContainer provides a place for many rows, so many that we might want to spill them into disk
```
```suggestion
// RowContainer provides a place for many rows, so many that we might want to spill them into disk.
```

SunRunAway commented on util/chunk/iterator.go:
```diff
@@ -284,3 +304,165 @@ func (it *iterator4RowPtr) ReachEnd() {
+		return it.End()
+	}
+	next := it.curIter.Next()
+	if next == it.curIter.End() {
```
We should do for loop here in case that the next iter is 0-length.

ichn-hu replied:
hahah, I've removed all iters that have 0 lenght

```
// NewMultiIterator creates a new multiIterator
func NewMultiIterator(iters ...Iterator) Iterator {
	iter := &multiIterator{}
	for i := 0; i < len(iters); i++ {
		if iters[i].Len() > 0 {
			iter.iters = append(iter.iters, iters[i])
			iter.length += iters[i].Len()
		}
	}
	iter.numIter = len(iter.iters)
	return iter
}
```

ichn-hu commented on executor/merge_join.go:
```diff
@@ -121,45 +121,50 @@ func (t *mergeJoinInnerTable) selectRow(row chunk.Row) {
 		t.curSel = nil
 	}
-	return t.sameKeyRows
+	return chunk.NewMultiIterator(iters...)
```
sounds good.
